### PR TITLE
fix(dto)!: correct some structural issues with the dtos and add tests

### PR DESCRIPTION
**Pull request by Alexandcoats**

### CODE REVIEW

**Review by grtlr — COMMENTED** [review]
We should use `bee-test` and add it to the `dev-dependencies`. `bee-test` can generate random variations of basically all the `bee-message` types.

File: src/types/stardust/message/address.rs [hunk 1]
```diff
@@ -24,15 +26,26 @@ impl TryFrom<Ed25519Address> for stardust::Ed25519Address {
...
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(stardust::Ed25519Address::from_str(s)?.into())
+    }
+}
+
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 pub enum Address {
+    #[serde(rename = "ed25519")]
     Ed25519(Ed25519Address),
+    #[serde(rename = "nft")]
```
[the remark this answers is not in the document]
Comment (Alexandcoats) [reply src/types/stardust/message/address.rs:41]: oop

File: src/types/stardust/message/output/alias.rs [hunk 2]
```diff
@@ -90,3 +106,66 @@ impl TryFrom<AliasOutput> for stardust::AliasOutput {
...
+    }
+
+    #[test]
+    fn test_alias_output_bson() {
+        let output = get_test_alias_output();
+        let bson = to_bson(&output).unwrap();
+        assert_eq!(output, from_bson::<AliasOutput>(bson).unwrap());
+    }
+
+    pub(crate) fn rand_alias_id() -> stardust::AliasId {
```
[the remark this answers is not in the document]
Comment (grtlr) [reply src/types/stardust/message/output/alias.rs:135]: Done: https://github.com/iotaledger/bee/pull/1409

So we will need to update `bee-message`.

File: src/types/stardust/message/output/basic.rs [hunk 3]
```diff
@@ -52,3 +52,54 @@ impl TryFrom<BasicOutput> for stardust::BasicOutput {
...
+    #[test]
+    fn test_basic_output_bson() {
+        let output = get_test_basic_output();
+        let bson = to_bson(&output).unwrap();
+        assert_eq!(output, from_bson::<BasicOutput>(bson).unwrap());
+    }
+
+    pub(crate) fn get_test_basic_output() -> BasicOutput {
+        BasicOutput::from(
+            &stardust::BasicOutput::build_with_amount(100)
```
Comment (grtlr) [inline src/types/stardust/message/output/basic.rs:79]: Prefix with `bee::`

File: src/types/stardust/message/output/basic.rs [hunk 4]
```diff
@@ -52,3 +52,54 @@ impl TryFrom<BasicOutput> for stardust::BasicOutput {
...
+    };
+
+    #[test]
+    fn test_basic_output_bson() {
+        let output = get_test_basic_output();
+        let bson = to_bson(&output).unwrap();
+        assert_eq!(output, from_bson::<BasicOutput>(bson).unwrap());
+    }
+
+    pub(crate) fn get_test_basic_output() -> BasicOutput {
```
Comment (grtlr) [inline src/types/stardust/message/output/basic.rs:77]: `bee_test::rand::output::rand_basic_output`

File: src/types/stardust/message/output/feature_block.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    pub(crate) fn get_test_issuer_block(address: Address) -> FeatureBlock {
+        FeatureBlock::Issuer { address }
+    }
+
+    pub(crate) fn get_test_metadata_block() -> FeatureBlock {
+        FeatureBlock::Metadata {
+            data: "Foo".as_bytes().to_vec().into_boxed_slice(),
+        }
```
Comment (grtlr) [inline src/types/stardust/message/output/feature_block.rs:102]: `bee_test::rand::output::*`

File: src/types/stardust/message/output/foundry.rs [hunk 6]
```diff
@@ -72,3 +72,45 @@ impl TryFrom<FoundryOutput> for stardust::FoundryOutput {
...
+    };
+
+    #[test]
+    fn test_foundry_output_bson() {
+        let output = get_test_foundry_output();
+        let bson = to_bson(&output).unwrap();
+        assert_eq!(output, from_bson::<FoundryOutput>(bson).unwrap());
+    }
+
+    pub(crate) fn get_test_foundry_output() -> FoundryOutput {
```
Comment (grtlr) [inline src/types/stardust/message/output/foundry.rs:93]: `bee_test`

File: src/types/stardust/message/output/mod.rs [hunk 7]
```diff
@@ -30,14 +32,14 @@ use crate::types::stardust::message::TransactionId;
...
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub struct OutputId {
     pub transaction_id: TransactionId,
     pub index: OutputIndex,
 }
 
-impl From<&stardust::OutputId> for OutputId {
-    fn from(value: &stardust::OutputId) -> Self {
+impl From<stardust::OutputId> for OutputId {
+    fn from(value: stardust::OutputId) -> Self {
```
Comment (grtlr) [inline src/types/stardust/message/output/mod.rs:42]: `bee::`

File: src/types/stardust/message/output/alias.rs [hunk 8]
```diff
@@ -1,15 +1,23 @@
...
 use super::{feature_block::FeatureBlock, native_token::NativeToken, unlock_condition::UnlockCondition, OutputAmount};
 
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 #[serde(transparent)]
 pub struct AliasId(#[serde(with = "serde_bytes")] pub Box<[u8]>);
 
+impl AliasId {
+    pub fn from_output_id_str(s: &str) -> Result<Self, crate::types::error::Error> {
+        Ok(stardust::AliasId::from(stardust::OutputId::from_str(s)?).into())
```
Comment (grtlr) [inline src/types/stardust/message/output/alias.rs:17]: `bee::`

File: src/types/stardust/message/output/native_token.rs [hunk 9]
```diff
@@ -45,7 +45,15 @@ impl TryFrom<TokenId> for stardust::TokenId {
     }
 }
 
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+impl FromStr for TokenId {
+    type Err = crate::types::error::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(stardust::TokenId::from_str(s)?.into())
```
Comment (grtlr) [inline src/types/stardust/message/output/native_token.rs:52]: `bee::`

File: src/types/stardust/message/output/nft.rs [hunk 10]
```diff
@@ -24,7 +32,15 @@ impl TryFrom<NftId> for stardust::NftId {
     }
 }
 
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+impl FromStr for NftId {
+    type Err = crate::types::error::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(stardust::NftId::from_str(s)?.into())
```
Comment (grtlr) [inline src/types/stardust/message/output/nft.rs:39]: `bee::`

File: src/types/stardust/message/output/unlock_condition.rs [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    #[serde(rename = "governor_address")]
+    GovernorAddress { address: Address },
+    #[serde(rename = "immutable_alias_address")]
+    ImmutableAliasAddress { alias_id: AliasId },
+}
+
+impl From<&stardust::UnlockCondition> for UnlockCondition {
+    fn from(value: &stardust::UnlockCondition) -> Self {
+        match value {
+            stardust::UnlockCondition::Address(a) => Self::Address {
```
Comment (grtlr) [inline src/types/stardust/message/output/unlock_condition.rs:40]: `bee::`

File: src/types/stardust/message/output/unlock_condition.rs [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+#[cfg(test)]
+pub(crate) mod test {
+    use mongodb::bson::{from_bson, to_bson};
+
+    use super::*;
+    use crate::types::stardust::message::output::alias::test::rand_alias_id;
+
+    #[test]
+    fn test_unlock_condition_bson() {
+        let block = get_test_address_condition(bee_test::rand::address::rand_address().into());
```
Comment (grtlr) [inline src/types/stardust/message/output/unlock_condition.rs:120]: `bee_test`

File: src/types/stardust/message/payload/milestone.rs [hunk 13]
```diff
@@ -24,9 +26,17 @@ impl TryFrom<MilestoneId> for stardust::MilestoneId {
     }
 }
 
+impl FromStr for MilestoneId {
+    type Err = crate::types::error::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(stardust::MilestoneId::from_str(s)?.into())
```
Comment (grtlr) [inline src/types/stardust/message/payload/milestone.rs:33]: `bee::`

File: src/types/stardust/message/payload/milestone.rs [hunk 14]
```diff
@@ -206,3 +216,115 @@ impl TryFrom<MigratedFundsEntry> for stardust::option::MigratedFundsEntry {
...
+
+    pub(crate) fn get_test_milestone_payload() -> MilestonePayload {
+        MilestonePayload::from(
+            &stardust::MilestonePayload::new(
+                get_test_milestone_essence().try_into().unwrap(),
+                vec![get_test_signature().try_into().unwrap()],
+            )
+            .unwrap(),
+        )
+    }
```
Comment (grtlr) [inline src/types/stardust/message/payload/milestone.rs:329]: `bee_test`

File: src/types/stardust/message/payload/mod.rs [hunk 15]
```diff
@@ -52,3 +52,47 @@ impl TryFrom<Payload> for stardust::Payload {
...
+
+    pub(crate) fn get_test_treasury_transaction_payload() -> Payload {
+        Payload::TreasuryTransaction(Box::new(
+            treasury_transaction::test::get_test_treasury_transaction_payload(),
+        ))
+    }
+
+    pub(crate) fn get_test_tagged_data_payload() -> Payload {
+        Payload::TaggedData(Box::new(tagged_data::test::get_test_tagged_data_payload()))
+    }
```
Comment (grtlr) [inline src/types/stardust/message/payload/mod.rs:97]: `bee_test`

**Review by grtlr — COMMENTED** [review]
I stopped commenting the `bee::` prefix and the `bee_test` remarks. :trollface:

File: bin/inx-chronicle/src/api/stardust/v2/routes.rs [hunk 16]
```diff
@@ -52,7 +51,7 @@ pub fn routes() -> Router {
 }
 
 async fn message(database: Extension<MongoDb>, Path(message_id): Path<String>) -> ApiResult<MessageResponse> {
-    let message_id_dto = MessageId::from(bee::MessageId::from_str(&message_id)?);
+    let message_id_dto = MessageId::from_str(&message_id).map_err(ParseError::BeeMessageStardust)?;
```
[the remark this answers is not in the document]
Comment (Alexandcoats) [reply bin/inx-chronicle/src/api/stardust/v2/routes.rs:54]: Well, actually we are, it's just hidden behind our FromStr impl.

File: src/types/stardust/message/output/alias.rs [hunk 17]
```diff
@@ -1,15 +1,23 @@
 // Copyright 2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
+use std::str::FromStr;
+
 use bee_message_stardust::output as stardust;
```
[the remark this answers is not in the document]
Comment (Alexandcoats) [reply src/types/stardust/message/output/alias.rs:6]: Something like `bee_message_stardust`? :trollface:

**Review by pvdrz — DISMISSED** [review]

**Pull request merged**